### ISSUE TITLE
PAST_THE_ERROR_MESSAGE_HERE

### ISSUE THREAD

**Issue by majorminors** [0x4447/0x4447_product_s3_email#36]
# Error Symptoms

CCed and BCCed emails get treated as 'Sent'.

# Notes

I decided to spin this up and have a play. It's wonderful!

I actually found it via a [thread on reddit](https://www.reddit.com/r/aws/comments/ci06pp/a_serverless_email_server_on_aws_using_s3_and_ses/). One comment asked about how bcc and cc were handled, and you specified that you weren't sure. I thought I'd test that. The poster's concerns were astute: items that you are cced or bcced in get treated as 'Sent' mail.

I wish I could have a crack at fixing this and submit a pull request, but I have only the vaguest idea of how this all works (the plan is to use your brilliant product to learn). I'll let you know if I eventually get there, but until then I thought I'd post the issue so you had it in the log.

Let me know if there's any outputs I can provide that would be useful for you to troubleshoot the issue.

**Comment by davidgatti** [comment]
Yes, that is true, and it is a bit annoying, but since I use this this to handle newsletters, and for signups email. it is not a big problem, sure some sites put emails addresses in the wrong places :) but o well. In any case, sure you can play with the code.

First of all this is a important link: https://github.com/topics/0x4447-product-s3-email - it shows all the resources of the project. And in this case you don't have to bother about CF, the code is in the separated Lambda Functions, that you can un locally.

And you should start looking at this line https://github.com/0x4447/0x4447_product_s3_email_lambda_inbound/blob/development/index.js#L207. 

Important, I won't accept code that is not formatted the way I formated it. Make sure to match the style and add meaningful comments.

**Comment by majorminors** [comment]
Can definitely see the point of leaving it, considering the use case of a signups/newsletters bin. We'll see if I can easily alter the code.

Cheers!

**Comment by davidgatti** [comment]
O I see so you would like to automatically forward every email to S3 Email, to have a backup?

**Comment by majorminors** [comment]
I was thinking to have an MX record pointing at the primary mailserver with, say, priority 10, and then a second MX record pointing at SES with, say, priority 20. So if the primary went down, SES and the 0x4447 s3 email stack would take over.

Alternatively, I could use SES as a relay for both use cases - so all email goes to SES. SES first tries to send it to my mail server. If that fails, SES uses the 0x4447 s3 email stack.

**Comment by davidgatti** [comment]
Hmmm impressive, I had no idea you could do that. And both ways are interesting, although the MX way seams to be simpler. Since in SES you'd have to make a rule and do some configuration. But yea in this case handling CC and BBC better would be good.

I think the best way would be to to have this flow:

1. To: is the most important field, and the email should be organized with that and ignore any other field
1. If To is not present, use CC
1. if CC is not present, use BCC

What do you think?

**Comment by majorminors** [comment]
I think that makes a lot of sense. Quick side note, just so I have it written down. I tested a TO as follows:

TO: someone_else@otherdomain.com; me@mydomain.com

This also gets treated as sent. I guess the filtering is just looking for the first domain it comes across in the TO field. I will also see if I can address this edge case, after the CC BCC thing.

**Comment by majorminors** [comment]
Would you prefer me to branch off the dev branch, or the master? I.e. how up to date are your dev branches? Reason I ask is that notice the commit history on the [incoming lambda repo](https://github.com/0x4447/0x4447_product_s3_email_lambda_inbound/tree/master) is different between the two--dev @ 74 vs master @ 86.

**Comment by davidgatti** [comment]
Will check tomorrow.

**Comment by majorminors** [comment]
No rush.

**Comment by davidgatti** [comment]
Sorry @majorminors, I was traveling, back to the main computer. Will check everything this weekend.

**Comment by davidgatti** [comment]
@majorminors Right if I merge from master to branch there are changes, but they are all PRs merges. If I check the network graph, all seams to be OK, meaning `development` is the main branch, and from there I did merge everything in to Master.

- https://github.com/0x4447/0x4447_product_s3_email_lambda_inbound/network
- https://github.com/0x4447/0x4447_product_s3_email_lambda_converter/network
- https://github.com/0x4447/0x4447_product_s3_email_lambda_outbound/network

So everything is OK? :) and if so, yes, you should branch out from `development` since it is the main one, and `master` is only for production. Meaning you branch out from dev, then make a PR that points your branch to dev, once that is OK, i merge dev to master.

**Comment by majorminors** [comment]
Hello!

My apologies for slow response this time. Didn't get a notification you'd responded.

Gotcha - my intuition was to branch from dev. Just wanted to check.

Assuming I can get this done, I'll be sure to PR your dev branch.

I'll keep you posted!

**Comment by davidgatti** [comment]
No worries, and cool :)

**Comment by majorminors** [comment]
Hello!

So the more I think about this, the more complicated I realise it could be if I try to account for all edge cases. The most problematic example would be if you had two or more SES verified domains (e.g. an email sent to two or more email addresses that you own) in any of the to, cc, or bcc fields.

I'm trying to keep to the basic philosophy you've specified in the readme and in the reddit thread. I.e. just a place to direct emails that aren't of critical importance.

As such, I wonder how you'd feel about:

1. First checks `to_domain` for an SES verified domain and if there's a hit, send it to the inbox (i.e. the current functionality) which as you say, will account for the vast majority of cases.
2. If there's no hit, checks to see if an SES verified domain exists anywhere else in the to, cc, or bcc fields. If there's a hit, then send it to the inbox under a path that is `uncategorised/from_domain/from_account`.
3. Otherwise, place in "sent".

The reason for using a pathname called "uncategorised" would be to indicate to the user that there's something weird about the email. Rather than lulling the user into a false sense of security that any edge cases would be taken care of by the script. It seems like this would be an improvement on having it default into the sent folder without going overboard.

The alternative seems like it would be to check each domain hit and try to regex the preceding string of each hit, followed by duplicating the email _n_ times to sort it accordingly, at which point one might as well pull a mail parser into the project. That feels like it goes against the grain of the project.

Let me know.

**Comment by davidgatti** [comment]
I will repeat what you wrote in my own way to see if we are on the same page:

So from the code, here I load all the domains added to SES: 

https://github.com/0x4447/0x4447_product_s3_email_lambda_inbound/blob/development/index.js#L116 

and then I go over what we have in the To field here: 

https://github.com/0x4447/0x4447_product_s3_email_lambda_inbound/blob/e74963a299879127a81682b0f41c4a54dcae607e/index.js#L409

to build out the location where to send. And what you are saying is that some emails might be directed at example.com based on the To field, but then in the CC or BCC fields the domain could be foo.com, where foo.com is your domain. And you are receiving the email form an external company and they are sending the same email to you the contractor for example. 

Is this the problem that you want to solve?

**Comment by majorminors** [comment]
Yes, this is the kind of edge case I was thinking about.

**Comment by davidgatti** [comment]
For sure there is a straight forward approach, and I think to figure it out, we can do the following: write down a bunch of examples of emails that cam be added to the "receiving" fields, and for each example the output that we want to see, this way based on that it will be easier to reason about and come up with a simple idea.

**Comment by loscil06** [comment]
You guys still working on this? Need any help for testing?

**Comment by davidgatti** [comment]
I don't think so.
